keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1338 @@
+	if err != nil {
+		return fmt.Errorf("failed to list images: %w", err)
+	}
+	c.loadImages(ctx, cImages)
+...
+...
+}
+</pre>
+
+
+**Proposed**
+
+<pre>
+
+// recover recovers system state from containerd and status checkpoint.
+func (c *criService) recover(ctx context.Context) error {
+...
+...
+	// Recover all images.
+    <mark>
+	listImagesOpt := []containerd.ListImageOpt {
+		containerd.ListImageWithPlatformMatcherMap(c.platformMatcherMap),
+	}
+	cImages, err := c.client.ListImagesWithPlatformMatcher(ctx, listImagesOpt...) </mark>
+	if err != nil {
+		return fmt.Errorf("failed to list images: %w", err)
+	}
+	c.loadImages(ctx, cImages)
+...
+...
+}
+
+
+<mark>
+// New functions added in client.go:
+</mark>
+
+func checkIfValidPlatforMatcher(ctx context.Context, provider content.Provider, image ocispec.Descriptor, matcher platforms.MatchComparer) bool {
+	_, err := images.Manifest(ctx, provider, image, matcher)
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+type ImagesWrapper struct {
+	Images Image
+	RuntimeHandler string
+}
+
+func (c *Client) ListImagesWithPlatformMatcher(ctx context.Context, opts ...ListImageOpt) ([]ImagesWrapper, error) {
+	var listImgOpts ListImageOptions
+	for _, o := range opts {
+		if err := o(ctx, &listImgOpts); err != nil {
+			return nil, err
+		}
+	}
+
+	imgs, err := c.ImageService().List(ctx, listImgOpts.Filters...)
+	if err != nil {
+		return nil, err
+	}
+
+	imagesW := make([]ImagesWrapper, len(imgs))
+	if len(opts) == 0 { 
+		for i, img := range imgs {
+			imagesW[i].Images = NewImage(c, img)
+			imagesW[i].RuntimeHandler = c.runtime
+		}
+	} else {
+		for i, img := range imgs {
+			for runtimeHandler, matcher := range listImgOpts.PlatformMatcherMap {
+				tmp := NewImage(c, img)
+				if checkIfValidPlatforMatcher(ctx, tmp.ContentStore(), tmp.Target(), matcher) {
+					imagesW[i].Images = NewImageWithPlatform(c, img, matcher)
+					imagesW[i].RuntimeHandler = runtimeHandler
+				}
+			}
+			if imagesW[i].Images == nil {
+				return nil, fmt.Errorf("valid platform matcher for image %v not found", img)
+			}
+		}
+	}
+	return imagesW, nil
+}
+
+</pre>
+
+
+### Windows platform matcher optional changes
+
+Windows platform matcher could optionally be extended to also check for host and UVM compatibility if hyperV isolation is being used based on the matrix described here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility 
+
+### Kubelet changes
+
+Kubelet needs to change how it talks about images from just the image name to (imageName, runtimeclass). As discussed in earlier sections, extending the Image struct in CRI to add a field for runtimeclass will make it easier to extend Kubelet API to include this information. 
+The following are scenarios where kubelet behavior needs to be extended to refer to image as its name and runtime class:
```

Comment:
Does kublet currently have access to the required information at every point when making the requests below?

Reply:
Yes. This is needed primarily:
1. EnsureImageExists() which is used to ensure an image exists before starting a pod or container. This has the needed info to identify the runtime class.
I see the same for RemoveImageRequest, ImageStatus and GC